src/lib/libc/time/mirtime.3
.\" $MirOS: src/lib/libc/time/mirtime.3,v 1.1 2011/11/20 04:57:12 tg Exp $
.\"-
.\" Copyright (c) 2004, 2005, 2007, 2011, 2018
.\"	mirabilos <m@mirbsd.org>
.\"
.\" Provided that these terms and disclaimer and all copyright notices
.\" are retained or reproduced in an accompanying document, permission
.\" is granted to deal in this work without restriction, including un-
.\" limited rights to use, publicly perform, distribute, sell, modify,
.\" merge, give away, or sublicence.
.\"
.\" This work is provided "AS IS" and WITHOUT WARRANTY of any kind, to
.\" the utmost extent permitted by applicable law, neither express nor
.\" implied; without malicious intent or gross negligence. In no event
.\" may a licensor, author or contributor be held liable for indirect,
.\" direct, other damage, loss, or other issues arising in any way out
.\" of dealing in the work, even if advised of the possibility of such
.\" damage or existence of a defect, except proven that it results out
.\" of said person's immediate fault when using the work as intended.
.\"-
.Dd $Mdocdate: November 20 2011 $
.ie \n(.g \{\
.	ds aq \(aq
.\}
.el \{\
.	ds aq '
.\}
.Dt MIRDATE 3
.Os
.Sh NAME
.Nm mirtime_getleaps ,
.Nm mirtime_isleap ,
.Nm timet2posix ,
.Nm posix2timet ,
.Nm timet2mjd ,
.Nm mjd2timet ,
.Nm timet2tm ,
.Nm tm2timet ,
.Nm mjd_explode ,
.Nm mjd_implode
.Nd MirBSD time API
.Sh SYNOPSIS
.Ft "const time_t *"
.Fn mirtime_getleaps "void"
.Ft "int"
.Fn mirtime_isleap "time_t t"
.Ft "time_t"
.Fn timet2posix "time_t t"
.Ft "time_t"
.Fn posix2timet "time_t t"
.Ft "mirtime_mjd *"
.Fn timet2mjd "mirtime_mjd *mjd" "time_t t"
.Ft "time_t"
.Fn mjd2timet "mirtime_mjd *mjd"
.Ft "struct tm *"
.Fn timet2tm "struct tm *dst" "time_t src"
.Ft "time_t"
.Fn tm2timet "struct tm *src"
.Ft "struct tm *"
.Fn mjd_explode "struct tm *tm" "mirtime_mjd *mjd"
.Ft "mirtime_mjd *"
.Fn mjd_implode "mirtime_mjd *mjd" "struct tm *tm"
.Sh DESCRIPTION
The
.Fn mirtime_getleaps
function returns a pointer, which is guaranteed to be not
.Dv NULL ,
to a read-only array of
.Vt time_t
values that are leap seconds, terminated by a value of 0.
This function uses a table which should be initialised with
.Fn tzset
beforehand if mechanisms such as
.Xr chroot 2
are employed.
.Pp
The
.Fn mirtime_isleap
function returns 1 if
.Fa t
is known to be a (positive) leap second, 0 otherwise.
.Pp
The remaining functions mentioned convert time values between
a broken-down calendar date
.Vt struct tm ,
a Modified Julian Date
.Vt mirtime_mjd ,
seconds since the First of January, 1970, midnight UTC
.Vt time_t ,
and POSIX time (seconds since the epoch, 86400 seconds per day),
also stored in
.Vt time_t .
.Pp
.Fn mjd_implode
and
.Fn tm2timet
parse the fields
.Ic tm_sec , tm_min , tm_hour , tm_mday ,
.Ic tm_mon , tm_year , tm_gmtoff
of
.Vt struct tm ;
.Fn mjd_explode
and
.Fn timet2tm
fill in the fields
.Ic tm_sec , tm_min , tm_hour , tm_mday ,
.Ic tm_mon , tm_year , tm_wday , tm_yday
with sensible values and
.Ic tm_isdst , tm_gmtoff , tm_zone
with null values.
.Pp
To use these functions, macros and types, include at least one of
.Pa \*(Ltsys/time.h\*(Gt
and/or
.Pa \*(Lttime.h\*(Gt
and check whether the macro
.Dv SYSKERN_MIRTIME_H
is defined
.Pq currently to the value 1 .
.Sh TYPES
.Vt time_t
is a signed integral type of 32 or 64 bit, machine-dependent.
.Pp
.Vt struct tm
consists of the following slots:
.Bd -literal -offset indent
int tm_sec;		/* seconds [0\-60] */
int tm_min;		/* minutes [0\-59] */
int tm_hour;		/* hours [0\-23] */
int tm_mday;		/* day of month [1\-31] */
int tm_mon;		/* month of year \- 1 [0\-11] */
time_t tm_year;		/* years since 1900 */
int tm_wday;		/* day of week (0 = sunday) */
int tm_yday;		/* day of year [0\-365] */
int tm_isdst;		/* summer time? [0 no, \*(Gt0 yes, \*(Lt0 unknown] */
long tm_gmtoff;		/* offset from UTC in seconds */
char *tm_zone;		/* abbreviation of timezone name */
.Ed
.Pp
.Vt mirtime_mjd
consists of the following slots:
.Bd -literal -offset indent
time_t mjd;		/* Julian Date minus 2\*(aq400\*(aq000.5, floor'd */
int32_t sec;		/* UTC (not TT) seconds into that [0\-86400] */
.Ed
.Sh SEE ALSO
.Xr chroot 2 ,
.Xr mktime 3 ,
.Xr tzset 3
.Sh STANDARDS
This set of functions expects your operating system to not conform to
.St -p1003.1
for correct
.Vt time_t
handling.
.Sh HISTORY
This API replaced a DJB-inspired TAI API with
.Mx 11 .
.Sh AUTHORS
.An Thorsten Do mirabilos Dc Glaser Aq tg@mirbsd.org
.Sh BUGS
The leap second table is read from the timezone information file.
This has implications on the location of the file and its up-to-dateness.
.Pp
There is no method to select POSIX-conformant behaviour.
It's probably better this way, though.
.Pp
Few sanitising of values is done, but the functions are considered safe.


src/kern/c/mirtime.c
/*-
 * Copyright © 2004, 2005, 2006, 2007, 2011
 *	Thorsten “mirabilos” Glaser <tg@mirbsd.org>
 *
 * Provided that these terms and disclaimer and all copyright notices
 * are retained or reproduced in an accompanying document, permission
 * is granted to deal in this work without restriction, including un‐
 * limited rights to use, publicly perform, distribute, sell, modify,
 * merge, give away, or sublicence.
 *
 * This work is provided “AS IS” and WITHOUT WARRANTY of any kind, to
 * the utmost extent permitted by applicable law, neither express nor
 * implied; without malicious intent or gross negligence. In no event
 * may a licensor, author or contributor be held liable for indirect,
 * direct, other damage, loss, or other issues arising in any way out
 * of dealing in the work, even if advised of the possibility of such
 * damage or existence of a defect, except proven that it results out
 * of said person’s immediate fault when using the work as intended.
 *-
 * MirBSD new time functions (common part, mirtime_getleaps is system
 * specific). The code to convert betwen MJD and broken-down calendar
 * dates is a re-implementation of algorithms from Dan J. Bernstein.
 *
 * This code works by the assumption that leap seconds are always po‐
 * sitive and happen at most once per cluster (no double leapsecs).
 */

#include <sys/types.h>
#include <syskern/mirtime.h>

__RCSID("$MirOS: src/kern/c/mirtime.c,v 1.2 2011/11/20 21:30:56 tg Exp $");

#ifdef L_timet2posix
time_t
timet2posix(time_t t)
{
	time_t u = t;
	const time_t *lp;

	if (__predict_true(t > 0)) {
		lp = mirtime_getleaps();
		while (__predict_true((*lp) && (t >= *lp))) {
			--u;
			++lp;
		}
	}
	return (u);
}
#endif

#ifdef L_posix2timet
time_t
posix2timet(time_t t)
{
	const time_t *lp;

	if (__predict_true(t > 0)) {
		lp = mirtime_getleaps();
		while (__predict_true((*lp) && (t >= *lp))) {
			++t;
			++lp;
		}
	}
	return (t);
}
#endif

#ifdef L_timet2mjd
mirtime_mjd *
timet2mjd(mirtime_mjd *mjd, time_t t)
{
	time_t u;

	/* POSIX time inherent property: days have 86400 seconds */
	u = timet2posix(t);
	mjd->sec = (int32_t)(u % 86400);
	u /= 86400;
	u += 40587;

	/* normalise */
	while (__predict_false(mjd->sec < 0)) {
		--u;
		mjd->sec += 86400;
	}

	/* account for a possible leap second */
	if (__predict_false(mirtime_isleap(t)))
		++mjd->sec;

	mjd->mjd = u;
	return (mjd);
}
#endif

#ifdef L_mjd2timet
time_t
mjd2timet(mirtime_mjd *mjd)
{
	time_t t;

	/* construct a POSIX timestamp again */
	t = mjd->mjd;
	t -= 40587;
	t *= 86400;
	t += (mjd->sec > 86399 ? 86399 : mjd->sec);

	/* add leap seconds between epoch and t */
	t = posix2timet(t);

	/* if mjd was a leap second, account for it */
	if (mjd->sec > 86399)
		++t;

	return (t);
}
#endif

#ifdef L_mjd_explode
struct tm *
mjd_explode(struct tm *tm, const mirtime_mjd *mjd)
{
	int sec, day, yday, mon;
	time_t year;

	year = mjd->mjd;
	sec = mjd->sec;

	/* normalise */
	while (sec < 0) {
		--year;
		sec += 86400;
	}
	while (sec > 86400) {
		++year;
		sec -= 86400;
	}

	/* buffer leap second */
	tm->tm_sec = (sec == 86400) ? 1 : 0;
	sec -= tm->tm_sec;

	/* calculate 400-year cycle (year) and offset in it (day) */
	day = (int)(year % 146097);
	year /= 146097;

	/* add bias: 678881 = days between "convenient origin" and MJD 0 */
	/* convenient origin is Wed(3) 1 March 0(fictional)/-1(real) */
	day += 678881;
	/* recalculate offset in cycle (Gregorian Period) */
	year += day / 146097;
	day %= 146097;

	/* days in 400 years are cyclic, they have 20871 weeks */
	tm->tm_wday = (day + 3) % 7;

	/* calculate year from period, taking leap years into account */
	year *= 4;
	/* a long (Julian) century is at the end of each Gregorian Period */
	if (__predict_false(day == 146096)) {
		year += 3;
		day = 36524;
	} else {
		year += day / 36524;
		day %= 36524;
	}
	year *= 25;
	year += day / 1461;
	day %= 1461;
	year *= 4;

	/* March to December, or January/February? */
	yday = (day < 306) ? 1 : 0;
	/* a leap year is at the end of each olympiad */
	if (__predict_false(day == 1460)) {
		year += 3;
		day = 365;
	} else {
		year += day / 365;
		day %= 365;
	}
	yday += day;

	/* count days and months from 1st March using fixed-point */
	day *= 10;
	mon = (day + 5) / 306;
	day = (day + 5) % 306;
	day /= 10;
	/* adjust for Jan/Feb offset */
	if (__predict_false(mon >= 10)) {
		mon -= 10;
		yday -= 306;
		++year;
	} else {
		mon += 2;
		yday += 31 + 28;
	}

	/* adjust for year 0(fictional) which did not exist */
	if (__predict_false(year < 1))
		--year;

	/* fill in the values still missing */
	tm->tm_sec += sec % 60;
	sec /= 60;
	tm->tm_min = sec % 60;
	tm->tm_hour = sec / 60;
	tm->tm_mday = day + 1;
	tm->tm_mon = mon;
	tm->tm_year = year - 1900;
	tm->tm_yday = yday;
	/* hardcode these */
	tm->tm_isdst = 0;
	tm->tm_gmtoff = 0;
	tm->tm_zone = (void *)0L;
	/* return dst */
	return (tm);
}
#endif

#ifdef L_mjd_implode
mirtime_mjd *
mjd_implode(mirtime_mjd *mjd, const struct tm *tm)
{
	time_t day;
	int x, y;

	/* get the seconds out first */
	mjd->sec = tm->tm_sec + 60 * tm->tm_min + 3600 * tm->tm_hour -
	    tm->tm_gmtoff;

	/* pull the year; adjust for year 0 not existing */
	if (__predict_false((day = tm->tm_year + 1900LL) < 0))
		++day;

	/* split year into y(ear in Gregorian Period) and convenient day */
	y = day % 400;
	day /= 400;
	day *= 146097;
	day -= 678882;
	/* add day of month */
	day += tm->tm_mday;

	/* normalise seconds into days */
	while (__predict_false(mjd->sec < 0)) {
		--day;
		mjd->sec += 86400;
	}
	while (__predict_false(mjd->sec > 86400)) {
		++day;
		mjd->sec -= 86400;
	}

	/* assign x the month and normalise into years */
	x = tm->tm_mon;
	while (__predict_false(x < 0)) {
		--y;
		x += 12;
	}
	y += x / 12;
	x %= 12;

	/* calculate x as month since march, adjust year for it */
	if (__predict_false(x < 2)) {
		x += 10;
		--y;
	} else
		x -= 2;

	/* add length of month since march by fixed-point arithmetic */
	day += (306 * x + 5) / 10;

	/* normalise 400-year cycles again */
	while (__predict_false(y < 0)) {
		day -= 146097;
		y += 400;
	}
	day += 146097 * (y / 400);
	y %= 400;

	/* add by year, 4 years, 100 years, 400 years */
	day += 365 * (y % 4);
	y /= 4;
	day += 1461 * (y % 25);
	day += 36524 * (y / 25);

	mjd->mjd = day;
	/* return dst */
	return (mjd);
}
#endif

#ifdef L_mirtime_isleap
int
mirtime_isleap(time_t t)
{
	time_t lt;
	const time_t *lp;

	lp = mirtime_getleaps();
	while (__predict_true((lt = *lp))) {
		if (__predict_false(t < lt))
			return (0);
		else if (__predict_false(t == lt))
			return (1);
		else
			++lp;
	}
	return (0);
}
#endif

#ifdef L_mirtime_macrobodies
#undef timet2tm
#undef tm2timet
struct tm *
timet2tm(struct tm *dst, time_t src)
{
	mirtime_mjd timet2tm_storage;

	return (mjd_explode(dst, timet2mjd(&timet2tm_storage, src)));
}

time_t
tm2timet(struct tm *src)
{
	mirtime_mjd tm2timet_storage;

	return (mjd2timet(mjd_implode(&tm2timet_storage, src)));
}
#endif
